drivers/ethernet/eth_numaker.c

```diff
@@ -0,0 +1,807 @@
+	key = k_spin_lock(&data->rx_frame_buf_lock);
+
+	/* Two approach: 1. recv all RX packets in one time.
+	 *               2. recv one RX and set pending interrupt for rx-next.
+	 */
+	while (1) {
+		/* get received frame */
+		if (m_numaker_gmacdev_get_rx_buf(gmacdev, &len, &buffer) != 0) {
+			break;
+		}
+
+		if (len == 0) {
+			break;
+		}
+		/* Allocate a memory buffer chain from buffer pool
+		 * Using root iface. It will be updated in net_recv_data()
+		 */
+		pkt = net_pkt_rx_alloc_with_buffer(data->iface, len, AF_UNSPEC, 0, K_NO_WAIT);
+		if (!pkt) {
+			LOG_ERR("pkt alloc frame-len=%d failed", len);
+			goto next;
+		}
+
+		LOG_DBG("length=%d, pkt=0x%x", len, (uint32_t)pkt);
+		/* deliver RX packet to upper layer, pack as one net_pkt */
+		if (net_pkt_write(pkt, buffer, len)) {
+			LOG_ERR("Unable to write RX frame into the pkt");
+			net_pkt_unref(pkt);
+			goto error;
+		}
+		if (pkt != NULL) {
+			int res = net_recv_data(data->iface, pkt);
+
+			if (res < 0) {
+				LOG_ERR("net_recv_data: %d", res);
+				net_pkt_unref(pkt);
+				goto error;
+			}
+		}
+next:
+		m_numaker_gmacdev_rx_next(gmacdev);
+	}
+	m_numaker_gmacdev_trigger_rx(gmacdev);
+
+error:
+	k_spin_unlock(&data->rx_frame_buf_lock, key);
+}
+
+static uint8_t *m_numaker_gmacdev_get_tx_buf(synopGMACdevice *gmacdev)
+{
+
+	DmaDesc *txdesc = gmacdev->TxNextDesc;
+
+	if (!synopGMAC_is_desc_empty(txdesc)) {
+		return NULL;
+	}
+
+	if (synopGMAC_is_desc_owned_by_dma(txdesc)) {
+		return NULL;
+	} else {
+		return (uint8_t *)(txdesc->buffer1);
+	}
+}
+
+static void m_numaker_gmacdev_trigger_tx(synopGMACdevice *gmacdev, uint16_t length)
+{
+	DmaDesc *txdesc = gmacdev->TxNextDesc;
+	uint32_t txnext = gmacdev->TxNext;
+#if defined(NU_USING_HW_CHECKSUM)
+	uint32_t offload_needed = 1;
+#else
+	uint32_t offload_needed = 0;
+#endif
+
+	/* busy tx descriptor is incremented by one as it will be handed over to DMA */
+	(gmacdev->BusyTxDesc)++;
+
+	txdesc->length |= ((length << DescSize1Shift) & DescSize1Mask);
+	txdesc->status |= (DescTxFirst | DescTxLast | DescTxIntEnable);
+
+	if (offload_needed) {
+		/*
+		 * Make sure that the OS you are running supports the IP and TCP checksum
+		 * offloading, before calling any of the functions given below.
+		 */
+		synopGMAC_tx_checksum_offload_tcp_pseudo(gmacdev, txdesc);
+	} else {
+		synopGMAC_tx_checksum_offload_bypass(gmacdev, txdesc);
+	}
+	__DSB();
+	txdesc->status |= DescOwnByDma;
+
+	gmacdev->TxNext = synopGMAC_is_last_tx_desc(gmacdev, txdesc) ? 0 : txnext + 1;
+	gmacdev->TxNextDesc =
+		synopGMAC_is_last_tx_desc(gmacdev, txdesc) ? gmacdev->TxDesc : (txdesc + 1);
+
+	/* Enable the interrupt */
+	synopGMAC_enable_interrupt(gmacdev, DmaIntEnable);
+	/* Trigger TX DMA */
+	synopGMAC_resume_dma_tx(gmacdev);
+}
+
+static int numaker_eth_tx(const struct device *dev, struct net_pkt *pkt)
+{
+	struct eth_numaker_data *data = dev->data;
+	synopGMACdevice *gmacdev = data->gmacdev;
+	uint16_t total_len = net_pkt_get_len(pkt);
+	uint8_t *buffer = NULL;
```

Comment:
no need to initialize this, it's set directly few lines below

Reply:
yes, buffer is no need to init.